flutter/ios/Classes/SwiftSentryFlutterPlugin.swift

```diff
@@ -3,27 +3,190 @@ import Sentry
+        if self.isValidSdk(sdk:sdk){
+
+            switch sdk["name"] as! String{
+            case "sentry.dart.flutter":
+                setEventEnvironmentTag(event:event,origin:"flutter",environment:"dart")
+            case "sentry.cocoa":
+                setEventEnvironmentTag(event:event,origin: "flutter",environment: "dart")
+            case "sentry.native":
+                setEventEnvironmentTag(event:event,origin: "flutter",environment: "dart")
+            default:
+                return
+            }
+        }
+    }
+
+    private func setEventEnvironmentTag(event: Event, origin: String = "ios", environment: String) {
+        event.tags?["event.origin"] = origin
+        event.tags?["event.environment"] = environment
+    }
+
+    private func isValidSdk( sdk: [String: Any]) -> Bool{
+        return (sdk["name"] != nil && !(sdk["name"] as! String).isEmpty)
+    }
+
+    func convertStringToDictionary(text: String) -> [String:AnyObject]? {
+        if let data = text.data(using: .utf8) {
+            do {
+                let json = try JSONSerialization.jsonObject(with: data, options: .mutableContainers) as? [String:AnyObject]
+                return json
+            } catch {
+                print("Something went wrong")
+            }
+        }
+        return nil
+    }
+
+    private func captureEnvelope(_ call: FlutterMethodCall, result: @escaping FlutterResult){
+        guard let arguments = call.arguments as? [Any],
+              !arguments.isEmpty,
+              let event = arguments.first as? String else {
+            result(FlutterError(code: "2", message: "Envelope is null or empty", details: nil) )
+            return
+        }
+
+
+        let parts = event.split(separator: "\n")
+        let envelopeParts: [[String: Any]] = try! parts.map ({ part in
+            convertStringToDictionary(text: "\(part)")!
+        })
+
+        guard let envelopeHeaderDict = envelopeParts[0] as? [String: Any],
+              let eventSdk =  envelopeHeaderDict as? [String: Any],
+              let eventId =  envelopeHeaderDict["event_id"] as? String,
+              let itemHeader = envelopeParts[1] as? [String: Any] ,
+              let itemType = itemHeader["type"] as? String,
+              let itemLength = itemHeader["length"] as? UInt else {
+            result(FlutterError(code: "2", message: "Cannot serialize event", details: nil) )
+            return
+        }
+
+        let sdkInfo = SentrySdkInfo(dict: eventSdk)
+        let sentryId = SentryId(uuidString: eventId)
+        let envelopeHeader = SentryEnvelopeHeader.init(id: sentryId, andSdkInfo: sdkInfo)
+
+        let sentryItemHeader = SentryEnvelopeItemHeader(type: itemType, length: itemLength)
+        
+        do {
+            let data = NSKeyedArchiver.archivedData(withRootObject: envelopeParts[2])
+            // TODO Fix Sentry - Error:: Failed to parse envelope item header Error Domain=NSCocoaErrorDomain Code=3840
```

Comment:
@marandaneto I don't understand why I get this parsing error. Do you have an idea?

Reply:
googling it I found this:

> Error Domain=NSCocoaErrorDomain Code=3840 "The operation couldn’t be completed. (Cocoa error 3840.)" (JSON text did not start with array or object and option to allow fragments not set.)

can we print each of the envelopeParts[0, 1 and 2] and see if they are valid JSONs or somehow broken?

Reply:
I've printed all the parts and everything seems fine. The SentryItemHeader is correctly instantiated, but it throws "internally" during the `client.capture(envelop)` call.
The SentryItemHeader is instantiated with json data converted to NsData, the internally it seems to re-converted to json, and then throws a parse error https://github.com/getsentry/sentry-cocoa/blob/e93d7de6a70202a6e4b5d405f0d77100220e76aa/Sources/Sentry/SentrySerialization.m#L161.

Reply:
@philipphofmann could you help us out with that?

Reply:
@rxlabz would you mind saving the envelope to a .txt file? thanks :)

Reply:
I seems like the JSON is not valid, like @marandaneto pointed it out. If you fix this and you get another error please ping me.

Reply:
<details>
<summary>header / itemHeader / payload</summary>
{"event_id":"5f941d3af7634830a2abc2fa32c557cd","sdk":{"name":"sentry.dart.flutter","version":"4.0.0-alpha.2","packages":[{"name":"pub:sentry","version":"4.0.0-alpha.2"},{"name":"pub:sentry_flutter","version":"4.0.0-alpha.2"}],"integrations":["isolateErrorIntegration","flutterErrorIntegration","nativeSdkIntegration","runZonedGuardedIntegration"]}}

{"content_type":"application/json","type":"event","length":5216}

{"event_id":"5f941d3af7634830a2abc2fa32c557cd","timestamp":"2020-11-17T08:31:18Z","platform":"other","release":"io.sentry.flutter.example@0.1.2+3","dist":"3","environment":"debug","exception":{"values":[{"type":"StateError","value":"Bad state: whats happening here","stacktrace":{"frames":[{"filename":"hooks.dart","function":"_dispatchPointerDataPacket","lineno":96,"colno":31,"abs_path":"dart:ui/hooks.dart","in_app":false},{"filename":"platform_dispatcher.dart","function":"PlatformDispatcher._dispatchPointerDataPacket","lineno":276,"colno":7,"abs_path":"dart:ui/platform_dispatcher.dart","in_app":false},{"filename":"hooks.dart","function":"_invoke1","lineno":176,"colno":10,"abs_path":"dart:ui/hooks.dart","in_app":false},{"filename":"zone.dart","function":"_CustomZone.runUnaryGuarded","lineno":1005,"colno":7,"abs_path":"dart:async/zone.dart","in_app":false},{"filename":"zone.dart","function":"_CustomZone.runUnary","lineno":1100,"colno":19,"abs_path":"dart:async/zone.dart","in_app":false},{"filename":"zone.dart","function":"_rootRunUnary","lineno":1206,"colno":13,"abs_path":"dart:async/zone.dart","in_app":false},{"filename":"binding.dart","package":"flutter","function":"GestureBinding._handlePointerDataPacket","lineno":221,"colno":7,"abs_path":"package:flutter/src/gestures/binding.dart","in_app":false},{"filename":"binding.dart","package":"flutter","function":"GestureBinding._flushPointerEventQueue","lineno":238,"colno":7,"abs_path":"package:flutter/src/gestures/binding.dart","in_app":false},{"filename":"binding.dart","package":"flutter","function":"GestureBinding.handlePointerEvent","lineno":280,"colno":5,"abs_path":"package:flutter/src/gestures/binding.dart","in_app":false},{"filename":"binding.dart","package":"flutter","function":"GestureBinding._handlePointerEventImmediately","lineno":316,"colno":7,"abs_path":"package:flutter/src/gestures/binding.dart","in_app":false},{"filename":"binding.dart","package":"flutter","function":"RendererBinding.dispatchEvent","lineno":279,"colno":11,"abs_path":"package:flutter/src/rendering/binding.dart","in_app":false},{"filename":"binding.dart","package":"flutter","function":"GestureBinding.dispatchEvent","lineno":361,"colno":22,"abs_path":"package:flutter/src/gestures/binding.dart","in_app":false},{"filename":"binding.dart","package":"flutter","function":"GestureBinding.handleEvent","lineno":381,"colno":19,"abs_path":"package:flutter/src/gestures/binding.dart","in_app":false},{"filename":"pointer_router.dart","package":"flutter","function":"PointerRouter.route","lineno":106,"colno":7,"abs_path":"package:flutter/src/gestures/pointer_router.dart","in_app":false},{"filename":"pointer_router.dart","package":"flutter","function":"PointerRouter._dispatchEventToRoutes","lineno":120,"colno":18,"abs_path":"package:flutter/src/gestures/pointer_router.dart","in_app":false},{"filename":"compact_hash.dart","function":"_LinkedHashMapMixin.forEach","lineno":377,"colno":8,"abs_path":"dart:collection-patch/compact_hash.dart","in_app":false},{"filename":"pointer_router.dart","package":"flutter","function":"PointerRouter._dispatchEventToRoutes.<fn>","lineno":122,"colno":9,"abs_path":"package:flutter/src/gestures/pointer_router.dart","in_app":false},{"filename":"pointer_router.dart","package":"flutter","function":"PointerRouter._dispatch","lineno":77,"colno":12,"abs_path":"package:flutter/src/gestures/pointer_router.dart","in_app":false},{"filename":"recognizer.dart","package":"flutter","function":"PrimaryPointerGestureRecognizer.handleEvent","lineno":476,"colno":9,"abs_path":"package:flutter/src/gestures/recognizer.dart","in_app":false},{"filename":"tap.dart","package":"flutter","function":"BaseTapGestureRecognizer.handlePrimaryPointer","lineno":222,"colno":7,"abs_path":"package:flutter/src/gestures/tap.dart","in_app":false},{"filename":"tap.dart","package":"flutter","function":"BaseTapGestureRecognizer._checkUp","lineno":287,"colno":5,"abs_path":"package:flutter/src/gestures/tap.dart","in_app":false},{"filename":"tap.dart","package":"flutter","function":"TapGestureRecognizer.handleTapUp","lineno":598,"colno":11,"abs_path":"package:flutter/src/gestures/tap.dart","in_app":false},{"filename":"recognizer.dart","package":"flutter","function":"GestureRecognizer.invokeCallback","lineno":183,"colno":24,"abs_path":"package:flutter/src/gestures/recognizer.dart","in_app":false},{"filename":"ink_well.dart","package":"flutter","function":"_InkResponseState._handleTap","lineno":990,"colno":20,"abs_path":"package:flutter/src/material/ink_well.dart","in_app":false},{"filename":"main.dart","package":"sentry_flutter_example","function":"_MyAppState.build.<fn>","lineno":81,"colno":32,"abs_path":"package:sentry_flutter_example/main.dart","in_app":true},{"filename":"main.dart","package":"sentry_flutter_example","function":"tryCatch","lineno":179,"colno":5,"abs_path":"package:sentry_flutter_example/main.dart","in_app":true}]}}]},"sdk":{"name":"sentry.dart.flutter","version":"4.0.0-alpha.2","packages":[{"name":"pub:sentry","version":"4.0.0-alpha.2"},{"name":"pub:sentry_flutter","version":"4.0.0-alpha.2"}],"integrations":["isolateErrorIntegration","flutterErrorIntegration","nativeSdkIntegration","runZonedGuardedIntegration"]}}
</details>

Here the code and the printed values :

```swift
// ...
print("itemType : \(itemType) , itemLength : \(itemLength)") // itemType : event , itemLength : 5216
let sentryItemHeader = SentryEnvelopeItemHeader(type: itemType, length: itemLength)
print("sentryItemHeader : \(sentryItemHeader)") // sentryItemHeader : <SentryEnvelopeItemHeader: 0x600002359200>

do {
  let data = NSKeyedArchiver.archivedData(withRootObject: envelopeParts[2])
  let sentryEnvelopeItem = SentryEnvelopeItem( header: sentryItemHeader, data: data)
  print("sentryEnvelopeItem \(sentryEnvelopeItem)") // sentryEnvelopeItem <SentryEnvelopeItem: 0x6000023a4060>
  let envelope = SentryEnvelope.init(header: envelopeHeader, singleItem: sentryEnvelopeItem)
  print("envelope \(envelope)") // envelope <SentryEnvelope: 0x6000023a40c0>

  SentrySDK.currentHub().getClient()?.capture(envelope: envelope)
  result("")
} catch {
  result(FlutterError(code: "2", message: "Cannot serialize event payload", details: nil) )
}
```

Reply:
JSON looks good @philipphofmann

Reply:
Ok it's fixed, I didn't pass the correct length.

Reply:
Ah okay, great to hear that @rxlabz. Thanks for the update 😀